### DISCUSSION TITLE
Is there an easy way to implement audit history log?

### DISCUSSION THREAD

**Discussion by taraktikos** [ent/ent#3165] [Q&A] [answered]
I am looking for something similar to https://www.baeldung.com/database-auditing-jpa

Currently, I am considering at least two possible solutions: 
1. To create a separate entity `Audit` and save data in the hook like it was described here https://entgo.io/docs/faq#how-to-write-an-audit-log-extension but in this case, we need to inject the client during code generation (I am not sure that ti is correct)
2. Another option is to create a read-only entity and use a trigger-based audit https://wiki.postgresql.org/wiki/Audit_trigger but there are also problems with trigger creation.

Could you please point me in the right direction? I also read about transactions hook but I am not sure how to implement it in a generic way

**Comment by crossworth** [comment, marked as answer]
Hello @taraktikos, this feature comes up a few times already on [Ent Discord](https://discord.gg/qZmPgTE6RX), there are a few links that I can provide you that may help you.

First, I would recommend that you check the [How to implement the Twitter edit button with Ent](https://entgo.io/blog/2022/04/07/how-twitter-can-implement/) blog post.

Second, I have on my projects a generic runtime hook that I use for storing audit logs, the way it works is more like storing changes, you can use as an example: https://gist.github.com/crossworth/8ef7e50e681ccf3f9d2ac79691d9c9d1
I store the audits on a `Audit` entity.

Maybe give a read on both links and if you still have any question, just ask and we can try make it work.

Oh, and feel free to join Ent discord.

**Reply by taraktikos** [reply]
Hello @crossworth. Thank you for your quick answer. It seems like this is what I was looking for
